gnovm/pkg/gnolang/preprocess.go

```diff
@@ -1701,13 +1701,23 @@ func Preprocess(store Store, ctx BlockNode, n Node) Node {
 
 			// TRANS_LEAVE -----------------------
 			case *ForStmt:
-				// Cond consts become bool *ConstExprs.
-				checkOrConvertType(store, last, &n.Cond, BoolType, false)
+				if nx, ok := (n.Cond).(*NameExpr); ok {
+					// Cond is named type, check if bool
+					checkNameExprType(store, last, nx, BoolType)
+				} else {
+					// Cond consts become bool *ConstExprs.
+					checkOrConvertType(store, last, &n.Cond, BoolType, false)
+				}
 
 			// TRANS_LEAVE -----------------------
 			case *IfStmt:
-				// Cond consts become bool *ConstExprs.
-				checkOrConvertType(store, last, &n.Cond, BoolType, false)
+				if nx, ok := (n.Cond).(*NameExpr); ok {
+					// Cond is named type, check if bool
+					checkNameExprType(store, last, nx, BoolType)
+				} else {
+					// Cond consts become bool *ConstExprs.
+					checkOrConvertType(store, last, &n.Cond, BoolType, false)
+				}
```

Comment:
I tend to think, actually, that if we change the original checkOrConvertType into [`checkOrConvertBoolType`](https://github.com/gnolang/gno/pull/2140/files#r1618434979) we can both optimise the processing and fix this issue.

Reply:
I think this should be the way to go.

Reply:
this should be the way to go.

Reply:
sounds good!